deployer/src/builder/oci/distribution/auth.rs

```diff
@@ -0,0 +1,141 @@
+use super::super::error::*;
+
+use serde::{Deserialize, Serialize};
+use std::{collections::HashMap, fs, io, path::*};
+use url::Url;
+
+/// Authentication info stored in filesystem
+#[derive(Debug, Clone, Default, Serialize, Deserialize)]
+pub struct StoredAuth {
+    auths: HashMap<String, Auth>,
+}
+
+impl StoredAuth {
+    /// Load authentication info with docker and podman setting
+    pub fn load_all() -> Result<Self> {
+        let mut auth = StoredAuth::default();
+        if let Some(path) = docker_auth_path() {
+            if let Ok(new) = Self::from_path(&path) {
+                auth.append(new)?;
+            }
+        }
+        if let Some(path) = podman_auth_path() {
+            if let Ok(new) = Self::from_path(&path) {
+                auth.append(new)?;
+            }
+        }
+        if let Some(path) = auth_path() {
+            let new = Self::from_path(&path)?;
+            auth.append(new)?;
+        }
+        Ok(auth)
+    }
+
+    /// Get token based on WWW-Authentication header
+    pub fn challenge(&self, challenge: &AuthChallenge) -> Result<String> {
+        let token_url = Url::parse(&challenge.url)?;
+        let domain = token_url
+            .domain()
+            .expect("www-authenticate header returns invalid URL");
+
+        let mut req = ureq::get(token_url.as_str()).set("Accept", "application/json");
+        if let Some(auth) = self.auths.get(domain) {
+            req = req.set("Authorization", &format!("Basic {}", auth.auth))
+        }
+        req = req
+            .query("scope", &challenge.scope)
+            .query("service", &challenge.service);
+        match req.call() {
+            Ok(res) => {
+                let token = res.into_json::<Token>()?;
+                Ok(token.token)
+            }
+            Err(ureq::Error::Status(..)) => Err(Error::AuthorizationFailed(token_url.to_string())),
+            Err(ureq::Error::Transport(e)) => Err(Error::Network(e.to_string())),
+        }
+    }
+
+    pub fn append(&mut self, other: Self) -> Result<()> {
+        for (key, value) in other.auths.into_iter() {
+            self.auths.insert(key, value);
+        }
```

Comment:
You might be able to use [`HashMap::extend()`](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.extend) here.